### PR TITLE
chore: implement P queries

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR implements P queries. See internal doc.

### CODE REVIEW

File: test/src/queryhandler.spec.ts [hunk 1]
```diff
@@ -351,4 +352,34 @@ describe('Query handler tests', function () {
       });
     });
   });
+
+  describe('P selectors', function () {
```
Comment (OrKoN) [inline test/src/queryhandler.spec.ts]: let's add tests with multiple query types including custom query handlers. Also, let's add coverage for the waitForSelector user case.

Maybe let's call it Puppeteer selectors in free-form texts?

Comment (jrandolf) [reply test/src/queryhandler.spec.ts]: > Also, let's add coverage for the waitForSelector user case.

This is already covered since P queries will be the default,

> Maybe let's call it Puppeteer Selectors in free-form texts?

I'm not sure. "Puppeteer selectors" is already an overloaded term.

We only need to add coverage for each of the query types and custom query handlers.

File: test/src/queryhandler.spec.ts [hunk 2]
```diff
@@ -351,4 +353,94 @@ describe('Query handler tests', function () {
...
+    it('should work with custom selectors', async () => {
+      const {page} = getTestState();
+      await page.setContent('<div>hello <button>world</button></div>');
+      Puppeteer.registerCustomQueryHandler('div', {
+        queryOne() {
+          return document.querySelector('div');
+        },
+      });
+
+      const element = await page.$(':-p-div()');
```
Comment (OrKoN) [inline test/src/queryhandler.spec.ts]: Let's add a test with a custom selector that has none empty parameters (perhaps once as a string value and once as an any value).

File: packages/puppeteer-core/src/common/util.ts [hunk 3]
```diff
@@ -488,3 +489,65 @@ export function stringifyFunction(expression: Function): string {
...
+ * @internal
+ */
+export async function* deepDescendents(
+  handle: ElementHandle<Node>
+): AwaitableIterable<ElementHandle<Node>> {
+  const walkers = await handle.evaluateHandle(element => {
+    return [document.createTreeWalker(element, NodeFilter.SHOW_ELEMENT)];
+  });
+  while (true) {
+    const handle = await walkers.evaluateHandle(walkers => {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/util.ts]: does it make the number of cdp calls to equal the number of shadow roots under an element? 

I don't think it's very efficient. Could we move the entirety of the search into the eval function to make it atomic and not so network heavy?

File: packages/puppeteer-core/src/common/PSelectorParser.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+export type PSelector =
+  | {
+      name: string;
+      value: string;
+    }
+  | CSSSelector;
+
+const PUPPETEER_FUNCTION_TOKEN = /^:-p-([-a-zA-Z_]+)\(/;
+
+class PSelectorParser {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/common/PSelectorParser.ts]: let's cover this with a unit-level tests? I think there a few edge cases that might be interesting and that are not covered by e2e.

File: packages/puppeteer-core/src/common/util.ts (hunk 3 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/util.ts]: This function is called whenever someone uses a `>>>` combinator. The number of CDP calls whenever used is equal to the number of elements in the elements tree excluding or included descendent trees (depending on the `deep*` function).

For querying a single element (which is most of the time), this is very efficient. It's also efficient if someone doesn't use `>>>`.

We should keep in mind that querying a shadow root is a very powerful, but expensive task. Since the largest use case is querying a single element, this is what we are optimizing for.

File: packages/puppeteer-core/src/common/PSelectorParser.ts (hunk 4 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/PSelectorParser.ts]: Could you provide these cases you are curious about? E2E should cover every case.

File: packages/puppeteer-core/src/common/util.ts (hunk 3 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/util.ts]: [EDIT] With batching included now, the number of CDP calls is equal to the number of elements / batch size.

**Comment by OrKoN** [comment]
@Lightning00Blade could you help out with the review once it's rebased?

File: packages/puppeteer-core/src/common/ExecutionContext.ts [hunk 5]
```diff
@@ -89,7 +91,9 @@ export class ExecutionContext {
     this._client = client;
     this._world = world;
     this._contextId = contextPayload.id;
-    this._contextName = contextPayload.name;
+    if (contextPayload.name) {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/ExecutionContext.ts:96]: Shouldn't this always exist?

File: test/TestExpectations.json [hunk 6]
```diff
@@ -1635,7 +1635,7 @@
     "testIdPattern": "",
     "platforms": ["darwin", "linux", "win32"],
     "parameters": ["webDriverBiDi"],
-    "expectations": ["SKIP", "TIMEOUT"]
+    "expectations": ["SKIP", "TIMEOUT", "FAIL"]
```
Comment (Lightning00Blade) [inline test/TestExpectations.json]: Why is this change made for BiDi?

File: packages/puppeteer-core/src/injected/PSelectorParser.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+  #scanParameter(): string {
+    const char = this.#input[0];
+    switch (char) {
+      case "'":
+      case '"':
+        this.#input = this.#input.slice(1);
+        const parameter = this.#scanEscapedValueTill(char);
+        if (!this.#input.startsWith(')')) {
+          throw new Error("Expected ')'");
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/injected/PSelectorParser.ts:119]: The error here are user facing right? Then we should give a better error messages.

File: packages/puppeteer-core/src/injected/PQuerySelector.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    selector = selector.trim();
+    if (selector.length === 0) {
+      throw new Error('The provided selector is empty.');
+    }
+    this.#deepShadowSelectors = parsePSelectors(selector);
+
+    // If there are any empty elements, then this implies the selector has
+    // contiguous combinators (e.g. `>>> >>>>`) or starts/ends with one which we
+    // treat as illegal, similar to existing behavior.
+    if (
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/injected/PQuerySelector.ts:46]: I think this should live inside the parser itself.

File: packages/puppeteer-core/src/injected/PQuerySelector.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+      }
+    }
+
+    for (; this.#selector !== undefined; this.#next()) {
+      const selector = this.#selector;
+      this.elements = AsyncIterableUtil.flatMap(
+        this.elements,
+        async function* (element) {
+          if (typeof selector === 'string') {
+            if (!element.parentElement) {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/injected/PQuerySelector.ts:95]: Nit:  inverse the `if`, imo it's easier to read

File: packages/puppeteer-core/src/common/ExecutionContext.ts (hunk 5 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/ExecutionContext.ts:96]: It may be an empty string which would imply it shouldn't.

File: test/TestExpectations.json (hunk 6 above)
Comment (jrandolf) [reply test/TestExpectations.json]: Because a test failed.

File: packages/puppeteer-core/src/injected/PQuerySelector.ts (hunk 9 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/injected/PQuerySelector.ts:95]: The current way it's set is the hot path.

File: packages/puppeteer-core/src/injected/PQuerySelector.ts (hunk 8 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/injected/PQuerySelector.ts:46]: The Parser doesn't check for semantically accurate selectors, just syntax.

File: test/TestExpectations.json (hunk 6 above)
Comment (OrKoN) [reply test/TestExpectations.json]: How can it fail if it is skipped? 🤔

**Review by OrKoN — COMMENTED** [review]
let's update the guides with the information about the new selector type https://pptr.dev/guides/query-selectors

**Comment by jrandolf** [comment]
> let's update the guides with the information about the new selector type https://pptr.dev/guides/query-selectors

Discussed offline.

**Review by OrKoN — APPROVED** [review]
